### PR TITLE
bpo-29986: Delete tip to raise TypeError from tp_richcompare.

### PR DESCRIPTION
**Pull request by ssbr**
I am not sure when `TypeError` is the right choice. Definitely, most of the time I've seen it done, it causes trouble, and `NotImplemented` usually does something better.

For example, see the work in https://bugs.python.org/issue8743 to get set to interoperate correctly with other set-like classes --- a problem caused by the use of `TypeError` instead of returning `NotImplemented` (e.g. https://hg.python.org/cpython/rev/3615cdb3b86d).

This advice seems to conflict with the usual and expected behavior of objects from Python: e.g. `object().__lt__(1)` returns `NotImplemented` rather than raising `TypeError`, despite `<` not "making sense" for `object`. Similarly for file objects and other uncomparable classes. Even complex numbers only return `NotImplemented`!

    >>> 1j.__lt__(1j)
    NotImplemented

OTOH, if this note should be kept, and this PR rejected, this section could use a decent explanation of the difference between "undefined" (should return `NotImplemented`) and "nonsensical" (should apparently raise `TypeError`). Perhaps a reference to an example from the stdlib.

https://bugs.python.org/issue29986

Automerge-Triggered-By: @tiran

### CODE REVIEW

**Comment by berkerpeksag** [comment]
Thanks for the PR, but please file an issue on bugs.python.org to discuss the patch.

**Comment by ssbr** [comment]
Done: http://bugs.python.org/issue29986

**Review by jdemeyer — APPROVED** [review]

File: Doc/c-api/typeobj.rst [hunk 1]
```diff
@@ -593,12 +593,6 @@ type objects) *must* have the :attr:`ob_size` field.
    ``Py_NotImplemented``, if another error occurred it must return ``NULL`` and
    set an exception condition.
 
-   .. note::
-
-      If you want to implement a type for which only a limited set of
-      comparisons makes sense (e.g. ``==`` and ``!=``, but not ``<`` and
-      friends), directly raise :exc:`TypeError` in the rich comparison function.
```
Comment (jdemeyer) [inline Doc/c-api/typeobj.rst:1318]: Instead of removing this note, maybe you should change it say return `NotImplemented` instead.

**Review by jdemeyer — CHANGES_REQUESTED** [review]

File: Doc/c-api/typeobj.rst (hunk 1 above)
Comment (brettcannon) [reply Doc/c-api/typeobj.rst:1318]: Unfortunately that clashes with guidance provided on the issue. :) Maybe this can get hashed out there?

Comment (jdemeyer) [reply Doc/c-api/typeobj.rst:1318]: > Unfortunately that clashes with guidance provided on the issue

The important thing is that the bad advice (`raise TypeError`) is removed. The consensus is to `return NotImplemented` in this case. How explicit you want to make that in the docs is bikeshedding territory.

In other words: I'm totally fine with just removing that paragraph.

**Comment by csabella** [comment]
@ericsnowcurrently, can you take a look at this change?  You recently modified this section of the docs (which is the merge conflict), but you didn't change this part, so your input may be helpful.  Thanks!

**Comment by ericsnowcurrently** [comment]
Unfortunately, I am not familiar enough right now with the relevant code to make a helpful review.  Sorry.

**Comment by ericsnowcurrently** [comment]
FWIW, to fix the merge conflict the relevant note just needs to be removed from wherever it is now in the file. :)

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by ssbr** [comment]
(I forgot about this PR; don't use github much and didn't notice the replies.) Is this blocked on me?

**Comment by JulienPalard** [comment]
Closing and reopening to retry failed CI job.

**Comment by JulienPalard** [comment]
@ssbr Would you please rebase this on master? The tests won't pass due to an expired SSL certificate (unrelated to your PR).

**Review by JulienPalard — APPROVED** [review]

**Comment by JulienPalard** [comment]
Hi @ssbr I rebased your branch on top of master because the CI was not passing due to someting completly unrelated. I had not the right to fix it on your branch (probably too old or me) so I reopened it for you here: https://github.com/python/cpython/pull/16095 and it should be merged any minute soon.

Thanks for your contribution!

**Pull request closed without merging**